config/v1/types_cluster_version.go

```diff
@@ -62,7 +62,7 @@ type ClusterVersionSpec struct {
 	//
 	// Some of the fields are inter-related with restrictions and meanings described here.
 	// 1. image is specified, version is specified, architecture is specified. API validation error.
-	// 2. image is specified, version is specified, architecture is not specified. You should not do this. version is silently ignored and image is used.
+	// 2. image is specified, version is specified, architecture is not specified. The version metadata in the referenced image must match the specified version.
```

Comment:
From the changes below about `Version`, is it more complete this way?
```suggestion
	// 2. image is specified, version is specified, architecture is not specified. image is used if the version metadata in the referenced image matches the specified version. API validation error otherwise.
```

Reply:
Lifted from Slack:

I think "API validation error" in the Godocs is trying to describe the [CEL](https://kubernetes.io/docs/reference/using-api/cel/) kubebuilder:validation:XValidation:rule [here](https://github.com/openshift/api/blob/965895d9d984b2d438784a7c483d5c3d200061c8/config/v1/types_cluster_version.go#L688-L689), because those are enforced by the Kube API server's validation, and clients cannot push invalid combinations.  For version, there's no CEL enforcement, it's just up to the CVO to decide how to handle the "image and version both set" situation, and report its thoughts in status.conditions

:TIL

Reply:
It needs to introspect the image, checking the version in [the release image's `release-metadata` file][1] against the `spec` `version` string that the cluster admin said they expected.  See [OCPBUGS-48641][2]'s:

> Verifying payload failed version="4.17.99" image="quay.io/openshift-release-dev/ocp-release@sha256:82aa2a914d4cd964deda28b99049abbd1415f96c0929667b0499dd968864a8dd" failure=release image version 4.17.13 does not match the expected upstream version 4.17.99

error message for an example, where the CVO looks inside the `sha256:82aa2a9...` release image, sees that the `release-metadata` file claims that release image is 4.17.13, and then complains that the release image's 4.17.13 diverges from the `spec` `version`'s 4.17.99 expectation.

[1]: https://github.com/openshift/enhancements/blob/4065a3c36352bec19fd144daa526e4c24fbd1aed/dev-guide/cluster-version-operator/user/reconciliation.md#release-image-content
[2]: https://issues.redhat.com/browse/OCPBUGS-48641

Reply:
Ack, and is this validation via a webhook or part of the controller? The `API validation error otherwise` message seems odd here, especially if this is a controller based validation

Reply:
it's via the controller (CVO).  And yes, that's why I would rather use my current text, and not the `API validation error` text.

Reply:
I would certainly be dropping the API validation error